File: 3980-best-time-to-buy-and-sell-stock-using-strategy/best-time-to-buy-and-sell-stock-using-strategy.java
class Solution {
    public long maxProfit(int[] prices, int[] strategy, int k) {
        int n=prices.length;
        long base_profit=0;
        for (int i=0;i<n;i++) {
            base_profit+=strategy[i]*prices[i];
        }
        long[] prefix_sum_prices=new long[n + 1];
        long[] prefix_sum_strategy=new long[n + 1];
        for(int i=0;i<n;i++) {
            prefix_sum_prices[i+1]=prefix_sum_prices[i]+prices[i];
            prefix_sum_strategy[i+1]=prefix_sum_strategy[i]+strategy[i]*prices[i];
        }
        long max_gain=0;
        int half=k/2;
        for (int i=0;i+k<=n;i++) {
            long old_segment=prefix_sum_strategy[i+k]-prefix_sum_strategy[i];
            long new_segment=prefix_sum_prices[i+k]-prefix_sum_prices[i+half];
            long gain=new_segment-old_segment;
            max_gain=Math.max(max_gain, gain);
        }
        return base_profit+Math.max(0,max_gain);
    }
}

File: 3980-best-time-to-buy-and-sell-stock-using-strategy/best-time-to-buy-and-sell-stock-using-strategy.cpp
class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        int n=prices.size();
        long long ans=0;
        vector<long long> org(n+1,0);
        for(int i=1;i<=n;i++){
            org[i]=org[i-1]+prices[i-1];
        }
         vector<long long> modify(n+1,0);
        for(int i=1;i<=n;i++){ 
            modify[i]=modify[i-1]+(prices[i-1]*strategy[i-1]);
        }
//         for(int i = 0; i <=n; i++) {
//     cout << org[i] << " ";
// }
// for(int i = 0; i <=n; i++) {
//     cout << modify[i] << " ";
// }
        int i=1;
        ans=modify[n];
        long long sum=0;
        //cout << '\n';
        n=n+1;
        while(i+k<=n){
            sum=0;
            sum+=modify[i-1];
           // cout << sum <<" ";
            sum+=(org[i+k-1]-org[i+k/2-1]);
           // cout << sum <<" ";
            sum+=(modify[n-1]-modify[i+k-1]);
            //cout << sum <<" ";
            ans=max(ans,sum);
            i++;
        }
        return ans;

    }
};